### ISSUE TITLE
Relax visbility rules for declaration emit

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#23127]
Visibility rules should be relaxed to only assert that a non-exported declaration is never exported.

```ts
class TreeNode {}
class Tree {
        // error TS4031: Public property 'Node' of exported class has or is using private name 'TreeNode'.
	Node: typeof TreeNode;
}
namespace Tree {
}
export = Tree;
```

another example:

```ts
interface I {
    a: number;
}

class C {
    // error TS4063: Parameter 'a' of constructor from exported class has or is using private name 'I'.
    constructor (a: I) { }
}

export default C;
```

**Comment by weswigham** [comment]
Is this about the error message?

**Comment by mhegazy** [comment]
We should not report the error in this case.. 
like today we ask is type of parameter `a` exported outside the module, and if no, we show an error.
What i am proposing is to change that to ask, does writing `I` make it visible, and if not show an error.

**Comment by weswigham** [comment]
Oh. So in other words,
```ts
class Tree {
        // error TS4031: Public property 'Node' of exported class has or is using private name 'TreeNode'.
	Node: typeof TreeNode;
}
```
causes us to also emit the 
```ts
class TreeNode {}
```
declaration (which would otherwise have been elided)? https://github.com/Microsoft/TypeScript/pull/22798 establishes a consistent framework for doing exactly that - I'd layer this on as another declaration kind added into that.
